firebase-installations/src/main/java/com/google/firebase/installations/remote/FirebaseInstallationServiceClient.java

```diff
@@ -0,0 +1,246 @@
+
+/** Http client that sends request to Firebase Installations backend API. */
+public class FirebaseInstallationServiceClient {
+  private static final String FIREBASE_INSTALLATIONS_API_DOMAIN =
+      "firebaseinstallations.googleapis.com";
+  private static final String CREATE_REQUEST_RESOURCE_NAME_FORMAT = "projects/%s/installations";
+  private static final String GENERATE_AUTH_TOKEN_REQUEST_RESOURCE_NAME_FORMAT =
+      "projects/%s/installations/%s/auth:generate";
+  private static final String DELETE_REQUEST_RESOURCE_NAME_FORMAT = "projects/%s/installations/%s";
+  private static final String FIREBASE_INSTALLATIONS_API_VERSION = "v1";
+  private static final String FIREBASE_INSTALLATION_AUTH_VERSION = "FIS_V2";
+
+  private static final String CONTENT_TYPE_HEADER_KEY = "Content-Type";
+  private static final String JSON_CONTENT_TYPE = "application/json";
+  private static final String CONTENT_ENCODING_HEADER_KEY = "Content-Encoding";
+  private static final String GZIP_CONTENT_ENCODING = "gzip";
+
+  public enum Code {
+    OK,
+
+    NETWORK_ERROR,
+
+    SERVER_ERROR,
+
+    UNAUTHORIZED,
+  }
+
+  @NonNull
+  public InstallationResponse createFirebaseInstallation(
+      long projectNumber,
+      @NonNull String apiKey,
+      @NonNull String firebaseInstallationId,
+      @NonNull String appId)
+      throws FirebaseException {
+    String resourceName = String.format(CREATE_REQUEST_RESOURCE_NAME_FORMAT, projectNumber);
+    try {
+      URL url =
+          new URL(
+              String.format(
+                  "https://%s/%s/%s?key=%s",
+                  FIREBASE_INSTALLATIONS_API_DOMAIN,
+                  FIREBASE_INSTALLATIONS_API_VERSION,
+                  resourceName,
+                  apiKey));
+
+      HttpsURLConnection httpsURLConnection = (HttpsURLConnection) url.openConnection();
+      httpsURLConnection.setDoOutput(true);
+      httpsURLConnection.setRequestMethod("POST");
+      httpsURLConnection.addRequestProperty(CONTENT_TYPE_HEADER_KEY, JSON_CONTENT_TYPE);
+      httpsURLConnection.addRequestProperty(CONTENT_ENCODING_HEADER_KEY, GZIP_CONTENT_ENCODING);
+      GZIPOutputStream gzipOutputStream =
+          new GZIPOutputStream(httpsURLConnection.getOutputStream());
+      try {
+        gzipOutputStream.write(
+            buildCreateFirebaseInstallationRequestBody(firebaseInstallationId, appId)
+                .toString()
+                .getBytes("UTF-8"));
+      } catch (JSONException e) {
+        throw new IllegalStateException(e);
+      } finally {
+        gzipOutputStream.close();
+      }
+
+      int httpResponseCode = httpsURLConnection.getResponseCode();
+      switch (httpResponseCode) {
+        case 200:
+          return readCreateResponse(httpsURLConnection);
+        case 401:
+          throw new FirebaseException("The request did not have the required credentials.");
+        default:
+          throw new FirebaseException("There was an internal server error.");
+      }
```

Comment:
Why does deleteFirebaseInstallation return a error code, but createFirebaseInstallation throws exceptions?

I think they should be consistent

Reply:
For example, both throw 
FirebaseInstallationNetworkException extends FirebaseException {
    Code code;
    FirebaseInstallationNetworkException(Code code, String message, ....) {
    }
}

Reply:
Done. Thanks.